### PR TITLE
Test with LuaRocks 3.1.3

### PR DESCRIPTION
**Pull request by blueyed**
Debugging issues with LuaRocks upgrade with Neovim (https://github.com/neovim/neovim/pull/10292).

### CODE REVIEW

File: appveyor.yml [hunk 1]
```diff
@@ -7,11 +7,18 @@ configuration:
...
+      NOCOMPAT: true  # with compatibility flags disabled.
+      LUAROCKS_VER: 2.3.0
+    - LJ_VER: 2.1
+      LUAROCKS_VER: 2.3.0
+
+    - LUA_VER: 5.3.2
+      NOCOMPAT: true  # with compatibility flags disabled.
+      LUAROCKS_VER: 3.1.3
+    - LJ_VER: 2.1
+      LUAROCKS_VER: 3.1.3
```
Comment (blueyed) [inline appveyor.yml]: Where / how is `LJ_VER` used?

**Comment by blueyed** [comment]
LuaRocks 3.1.3 installation fails on Travis:
```
[1;36mConfiguring LuaRocks version 3.1.3...[0m

Lua version detected: [1;32m5.1[0m
Lua interpreter found: [1;32m/home/travis/.lua/luajit[0m
lua.h found: [1;32m/home/travis/build/luvit/luv/install/lua/include/luajit-2.1/lua.h[0m
unzip found in PATH: [1;32m/usr/bin[0m

[1;36mDone configuring.[0m

LuaRocks will be installed at......: [1;32m/home/travis/build/luvit/luv/install/luarocks[0m
LuaRocks will install rocks at.....: [1;32m/home/travis/build/luvit/luv/install/luarocks[0m
LuaRocks configuration directory...: [1;32m/home/travis/build/luvit/luv/install/luarocks/etc/luarocks[0m
Using Lua from.....................: [1;32m/home/travis[0m
Lua include directory..............: [1;32m/home/travis/build/luvit/luv/install/lua/include/luajit-2.1[0m

* Type [1mmake[0m and [1mmake install[0m:
  to install to /home/travis/build/luvit/luv/install/luarocks as usual.
* Type [1mmake bootstrap[0m:
  to install LuaRocks into /home/travis/build/luvit/luv/install/luarocks as a rock.

mkdir -p "build"
rm -f src/luarocks/core/hardcoded.lua
echo "#!/bin/sh" > luarocks
echo "unset LUA_PATH LUA_PATH_5_2 LUA_PATH_5_3 LUA_PATH_5_4" >> luarocks
echo 'LUAROCKS_SYSCONFDIR="/home/travis/build/luvit/luv/install/luarocks/etc/luarocks" LUA_PATH="/home/travis/build/luvit/luv/luarocks-3.1.3/src/?.lua;;" exec "/home/travis/.lua/luajit" "/home/travis/build/luvit/luv/luarocks-3.1.3/src/bin/luarocks" --project-tree="/home/travis/build/luvit/luv/luarocks-3.1.3/lua_modules" "$@"' >> luarocks
chmod +rx ./luarocks
./luarocks init

Initializing project 'luarocks-3.1.3' for Lua 5.1 ...
-----------------------------------------------------

Checking your Lua installation ...

Error: Failed finding Lua header files. You may need to install them or configure LUA_INCDIR.
make: *** [luarocks] Error 1
.ci/setup_lua.sh: line 110: luarocks: command not found
.ci/setenv_lua.sh: line 3: /home/travis/.lua/luarocks: No such file or directory
```

**Comment by blueyed** [comment]
Would appreciate some pointer(s) here - I think it is good/required to test with LuaRocks 3 also.

**Comment by squeek502** [comment]
Looks good to me, but it's currently blocked by a Luarocks bug, right? Maybe we should allow Luarocks 3.1.3 jobs to fail for now?
- https://docs.travis-ci.com/user/customizing-the-build#rows-that-are-allowed-to-fail
- https://www.appveyor.com/docs/build-configuration/#allow-failing-jobs

Also worth noting that this will balloon the time it takes for Appveyor to complete--after this PR, it will likely take an hour. Might want to cut that down a bit?

**Comment by blueyed** [comment]
> Looks good to me, but it's currently blocked by a Luarocks bug, right?

I'm not sure.
Found some issue(s) with it, but also in combination with luv (mostly), therefore I've thought it could be due to wrong invocation.
On Linux: https://github.com/luarocks/luarocks/issues/950#issuecomment-504741854

The Neovim WIP PR to upgrade LuaRocks: https://github.com/neovim/neovim/pull/10292

> Maybe we should allow Luarocks 3.1.3 jobs to fail for now?

I'd rather wait for it to be fixed then / keep this WIP.

Cutting jobs down before merging seems sensible, yes.

**Comment by squeek502** [comment]
Appveyor CI should be able to be fixed by changing this line:
https://github.com/luvit/luv/blob/b901b0046bc153ff1e7a14a9acbc2f6734680aeb/.ci/install.bat#L28

to
```
if not defined LUAROCKS_URL set LUAROCKS_URL=https://luarocks.github.io/luarocks/releases
```

**Comment by blueyed** [comment]
Should be good on Travis now, AppVeyor still failed.

Maybe it requires changes to the install.bat call, e.g. `/CONFIG` (https://github.com/luarocks/luarocks/issues/1046)?
Are you able to test this locally (i.e. do you have a Windows machine)?
Otherwise feel free to push fixups here.

**Comment by squeek502** [comment]
`/CONFIG` did the trick but https://github.com/luvit/luv/commit/9d28d38cfd0b6d966c061a35cf89b930364e228b is also needed. Will rebase this PR on latest master in a sec.

**Comment by blueyed** [comment]
LuaRocks fails to find LuaJIT headers now:
```
./luarocks init
Initializing project 'luarocks-3.1.3' for Lua 5.1 ...
-----------------------------------------------------
Checking your Lua installation ...
Error: Failed finding Lua header files. You may need to install them or configure LUA_INCDIR.
make: *** [luarocks] Error 1
+ln -s /Users/travis/build/luvit/luv/install/luarocks/bin/luarocks /Users/travis/.lua/luarocks
+cd /Users/travis/build/luvit/luv
+luarocks --version
.ci/setup_lua.sh: line 122: luarocks: command not found
```

1. the above shoutl use `set -e` to abort already.
2. there should be LuaJIT 2.0 in the job matrix also (I assume it might work there).
3. Likely more an issue with LuaRocks? (or the configure args)  Ref: https://github.com/neovim/neovim/pull/10321/files#diff-4827d9465b6d6428c1ce9ca4b814cb69

**Comment by squeek502** [comment]
LuaRocks 3.1.3 + LuaJIT 2.1 error is related to https://github.com/luarocks/luarocks/pull/1042 / https://github.com/luarocks/luarocks/pull/1051, right?

**Comment by blueyed** [comment]
Not necessarily - https://github.com/luarocks/luarocks/pull/1051 does not really use different include dirs, but only helps with LuaJIT vs Lua (from a quick glance).

**Comment by squeek502** [comment]
Not sure what's going wrong, but Lua 5.3 works while LuaJIT 2.1 doesn't.

From the [failing LuaRocks 3.1.3 + LuaJIT 2.1 build](https://travis-ci.org/luvit/luv/jobs/559889286):
```
./configure --prefix=/home/travis/build/luvit/luv/install/luarocks --with-lua-interpreter=luajit --with-lua-include=/home/travis/build/luvit/luv/install/lua/include/luajit-2.1

Configuring LuaRocks version 3.1.3...

Lua version detected: 5.1
Lua interpreter found: /home/travis/.lua/luajit
lua.h found: /home/travis/build/luvit/luv/install/lua/include/luajit-2.1/lua.h
unzip found in PATH: /usr/bin

Done configuring.

LuaRocks will be installed at......: /home/travis/build/luvit/luv/install/luarocks
LuaRocks will install rocks at.....: /home/travis/build/luvit/luv/install/luarocks
LuaRocks configuration directory...: /home/travis/build/luvit/luv/install/luarocks/etc/luarocks
Using Lua from.....................: /home/travis
Lua include directory..............: /home/travis/build/luvit/luv/install/lua/include/luajit-2.1
```

The Lua include directory there is correct (it matches the `lua.h found` above).

From the [passing LuaRocks 3.1.3 + Lua 5.3 build](https://travis-ci.org/luvit/luv/jobs/559889285):
```
./configure --with-lua=/home/travis/build/luvit/luv/install/lua --prefix=/home/travis/build/luvit/luv/install/luarocks

Configuring LuaRocks version 3.1.3...

Lua version detected: 5.3
Lua interpreter found: /home/travis/build/luvit/luv/install/lua/bin/lua
lua.h found: /home/travis/build/luvit/luv/install/lua/include/lua.h
unzip found in PATH: /usr/bin

Done configuring.

LuaRocks will be installed at......: /home/travis/build/luvit/luv/install/luarocks
LuaRocks will install rocks at.....: /home/travis/build/luvit/luv/install/luarocks
LuaRocks configuration directory...: /home/travis/build/luvit/luv/install/luarocks/etc/luarocks
Using Lua from.....................: /home/travis/build/luvit/luv/install/lua
```

**Comment by blueyed** [comment]
Looks correct, yes.

The error is coming from https://github.com/luarocks/luarocks/blob/fe9af86aaad5aff6ceacd9cfd2d4e9438e12c40e/src/luarocks/deps.lua#L560, so it might indeed be related to the luajit_version (fixed/simplified in https://github.com/luarocks/luarocks/pull/1051).
Should we test that branch here?

**Comment by squeek502** [comment]
Doesn't fix it:
- https://github.com/squeek502/luv/commit/67d8ba5b60a9bccd099b445c2cc372653073accc
- https://travis-ci.org/squeek502/luv/jobs/560218136

**Comment by blueyed** [comment]
```
+./configure --prefix=/home/travis/build/squeek502/luv/install/luarocks --with-lua-interpreter=luajit --with-lua-include=/home/travis/build/squeek502/luv/install/lua/include/luajit-2.1
Configuring LuaRocks version dev...
Lua version detected: 5.1
Lua interpreter found: /home/travis/.lua/luajit
lua.h found: /home/travis/build/squeek502/luv/install/lua/include/luajit-2.1/lua.h
unzip found in PATH: /usr/bin
Done configuring.
LuaRocks will be installed at......: /home/travis/build/squeek502/luv/install/luarocks
LuaRocks will install rocks at.....: /home/travis/build/squeek502/luv/install/luarocks
LuaRocks configuration directory...: /home/travis/build/squeek502/luv/install/luarocks/etc/luarocks
Using Lua from.....................: /home/travis
Lua include directory..............: /home/travis/build/squeek502/luv/install/lua/include/luajit-2.1
* Type make and make install:
  to install to /home/travis/build/squeek502/luv/install/luarocks as usual.
* Type make bootstrap:
  to install LuaRocks into /home/travis/build/squeek502/luv/install/luarocks as a rock.
+make build
mkdir -p "build"
rm -f src/luarocks/core/hardcoded.lua
echo "#!/bin/sh" > luarocks
echo "unset LUA_PATH LUA_PATH_5_2 LUA_PATH_5_3 LUA_PATH_5_4" >> luarocks
echo 'LUAROCKS_SYSCONFDIR="/home/travis/build/squeek502/luv/install/luarocks/etc/luarocks" LUA_PATH="/home/travis/build/squeek502/luv/luarocks-dynamic-luajit/src/?.lua;;" exec "/home/travis/.lua/luajit" "/home/travis/build/squeek502/luv/luarocks-dynamic-luajit/src/bin/luarocks" --project-tree="/home/travis/build/squeek502/luv/luarocks-dynamic-luajit/lua_modules" "$@"' >> luarocks
chmod +rx ./luarocks
./luarocks init
Error: Failed finding Lua header files. You may need to install them or configure LUA_INCDIR.
```

Can you try "make bootstrap" instead of "luarocks init"?
Also dumping `luarocks config` might be useful.

**Comment by squeek502** [comment]
That `./luarocks init` is part of [`make build`](https://github.com/luarocks/luarocks/blob/526664e6e54da0f129621ea6df355767259a103a/GNUmakefile#L55)

**Comment by hishamhm** [comment]
@blueyed @squeek502 I finally managed to look into this. The build failure [here](https://travis-ci.org/squeek502/luv/jobs/560218136) happens because the `configure` script gets confused by the `~/.lua/lua` symlink when trying to find the interpreter, then assumes the prefix for Lua is `~`, which is wrong. Since autodetection is best-effort and the prefix of the Lua/LuaJIT install is known, what I recommend is using `--with-lua=` and telling the `configure` script explicitly. 

I tested the suggested change here:

* https://github.com/hishamhm/luv/commit/09a29f072c5063134e333bacf96f89987f741d96

And got a working build: 

* :green_heart: https://travis-ci.org/hishamhm/luv/builds/564118010

(Thank you @squeek502 for providing an easy-to-test environment in your fork!)

**Comment by squeek502** [comment]
Thanks for testing and finding the problem, @hishamhm! Out of curiosity, I just tested and confirmed that the fix actually works with stock Luarocks 3.1.3:

- https://travis-ci.org/squeek502/luv/builds/564189616
- https://github.com/squeek502/luv/commits/test-luarocks-test

(i.e. https://github.com/luarocks/luarocks/pull/1051 isn't needed in this case).

**Comment by blueyed** [comment]
> I think this is good to go.

Yay!  Should be either squash-merged then, or cleaned up manually and force-pushed.

> Appveyor CI takes ~50 minutes to complete now, but that can be addressed in a future PR.

How would you address it there then?
Maybe it can be pruned already here?  (https://github.com/luvit/luv/pull/344/files#diff-180360612c6b8c4ed830919bbb4dd459)

**Comment by hishamhm** [comment]
@squeek502 Thank you for the feedback! I also further tweaked my PR so that your original code from https://github.com/squeek502/luv/commit/67d8ba5b60a9bccd099b445c2cc372653073accc passes when using LuaRocks from the PR's branch:

https://travis-ci.org/hishamhm/luv/builds/564220755

**Comment by squeek502** [comment]
> How would you address it there then?

Not sure yet, which is why I was thinking of doing it separately. Options I was thinking about:

- cut out the MSVC 2013 CI builds entirely or just leave one pair of x86/x64 for MSVC 2013
- try to cut down on the number of jobs by removing probably-redundant jobs (like only testing one pair of x86/x64 and then skipping x86 for the rest)

Both have their potential drawbacks, which is another reason why it might be good to do it separately so it can be discussed more on its own.

**Pull request merged**